File: src/fable-library-py/fable_library/task_builder.py
from __future__ import annotations

from collections.abc import Awaitable, Callable, Iterable
from typing import (
    Any,
    Protocol,
    overload,
)

from .task import from_result, zero
from .util import IDisposable


class Delayed[T](Protocol):
    def __call__(self, __unit: None | None = None) -> Awaitable[T]: ...


class TaskBuilder:
    async def Bind[T, U](self, computation: Awaitable[T], binder: Callable[[T], Awaitable[U]]) -> U:
        async def bind() -> U:
            value = await computation
            return await binder(value)

        return await bind()

    async def Combine[T](self, computation1: Awaitable[None], computation2: Delayed[T]) -> T:
        return await self.Bind(computation1, computation2)

    def Delay[T](self, generator: Callable[[], Awaitable[T]]) -> Delayed[T]:
        def deferred(_: Any = None) -> Awaitable[T]:
            # print("Delay: deferred: ", generator)
            return generator()

        return deferred

    async def For[T](self, sequence: Iterable[T], body: Callable[[T], Awaitable[None]]) -> None:
        done = False
        it = iter(sequence)
        try:
            cur = next(it)
        except StopIteration:
            done = True

        def delay():
            nonlocal cur, done
            res = body(cur)
            try:
                cur = next(it)
            except StopIteration:
                done = True
            return res

        return await self.While(lambda: not done, self.Delay(delay))

    @overload
    async def Return(self) -> None: ...

    @overload
    async def Return[T](self, value: T) -> T: ...

    async def Return(self, value: Any = None) -> Any:
        return await from_result(value)

    async def ReturnFrom[T](self, computation: Awaitable[T]) -> T:
        return await computation

    async def TryFinally[T](self, computation: Delayed[T], compensation: Callable[[], None]) -> T:
        async def try_finally() -> T:
            try:
                t = await computation()
            finally:
                compensation()
            return t

        return await try_finally()

    async def TryWith[T](self, computation: Delayed[T], catchHandler: Callable[[Any], Awaitable[T]]) -> T:
        async def try_with() -> T:
            try:
                t = await computation()
            except Exception as exn:
                t = await catchHandler(exn)
            return t

        return await try_with()

    async def Using[T: IDisposable, U](self, resource: T, binder: Callable[[T], Awaitable[U]]) -> U:
        return await self.TryFinally(self.Delay(lambda: binder(resource)), lambda: resource.Dispose())

    @overload
    async def While(self, guard: Callable[[], bool], computation: Delayed[None]) -> None: ...

    @overload
    async def While[T](self, guard: Callable[[], bool], computation: Delayed[T]) -> T: ...

    async def While(self, guard: Callable[[], bool], computation: Delayed[Any]) -> Any:
        if guard():
            return await self.Bind(computation(), lambda _: self.While(guard, computation))
        else:
            return await self.Return()

    async def Zero(self) -> None:
        return await zero()

    async def Run[T](self, computation: Delayed[T]) -> T:
        # Make sure we don't execute computation right now, so wrap in a coroutine.
        async def run() -> T:
            return await computation()

        return await run()


task = TaskBuilder

__all__ = ["task"]

File: src/fable-library-ts/TaskBuilder.ts
import { IDisposable } from "./Util.ts";
import { zero } from "./Task.ts";

export class TaskBuilder {
  public Bind<T, U>(computation: Promise<T>, binder: (x: T) => Promise<U>): Promise<U> {
    return computation.then(binder);
  }

  public Combine<T>(computation1: Promise<void>, computation2: () => Promise<T>): Promise<T> {
    return computation1.then(computation2);
  }

  public Delay<T>(generator: () => Promise<T>): () => Promise<T> {
    return generator;
  }

  public For<T>(sequence: Iterable<T>, body: (x: T) => Promise<void>): Promise<void> {
    const iter = sequence[Symbol.iterator]();
    let cur = iter.next();
    return this.While(
      () => !cur.done,
      this.Delay(() => {
        const res = body(cur.value!);
        cur = iter.next();
        return res;
      })
    );
  }

  public Return<T>(value?: T): Promise<T | undefined> {
    return Promise.resolve(value);
  }

  public ReturnFrom<T>(computation: Promise<T>): Promise<T> {
    return computation;
  }

  public TryFinally<T>(computation: () => Promise<T>, compensation: () => void): Promise<T> {
    try {
      return computation().finally(compensation);
    } catch (e) {
      compensation();
      return Promise.reject(e);
    }
  }

  public TryWith<T>(computation: () => Promise<T>, catchHandler: (e: unknown) => Promise<T>): Promise<T> {
    try {
      return computation().catch(catchHandler);
    } catch (e) {
      try {
        return catchHandler(e);
      } catch (e2) {
        return Promise.reject(e2);
      }
    }
  }

  public Using<T extends IDisposable, U>(resource: T, binder: (x: T) => Promise<U>): Promise<U> {
    return this.TryFinally(() => binder(resource), () => resource.Dispose());
  }

  public While(guard: () => boolean, computation: () => Promise<void>): Promise<void> {
    return (async () => {
      while (guard()) {
        await computation();
      }
    })();
  }

  public Zero(): Promise<void> {
    return zero();
  }

  public Run<T>(computation: () => Promise<T>): Promise<T> {
    try {
      return computation();
    } catch (e) {
      return Promise.reject(e);
    }
  }
}

export const singleton = new TaskBuilder();
export function task(): TaskBuilder { return singleton; }
